keps/sig-api-machinery/4568-resilient-watchcache-initialization/README.md

```diff
@@ -0,0 +1,464 @@
+The way that PostStartHook work is that it is started exacty [once on startup] and is
+[hanging until success]. As soon as it succeeds it [starts reporting success] and never
+changes until kube-apiserver stops. We use that mechanism (instead of regular readyz check)
+because watchache is per-type layer and we want to avoid marking the whole kube-apiserver
+as `not-ready` if watchcache for one of the resource types requires reinitialization,
+because requests for all other resource types can still be handled properly.
+To handle those cases, we will use a different mechanisms described below (rejection).
+
+[once on startup]: https://github.com/kubernetes/kubernetes/blob/release-1.30/staging/src/k8s.io/apiserver/pkg/server/hooks.go#L155-L168
+[hanging until success]: https://github.com/kubernetes/kubernetes/blob/release-1.30/staging/src/k8s.io/apiserver/pkg/server/hooks.go#L194-L206
+[starts reporting success]: https://github.com/kubernetes/kubernetes/blob/release-1.30/staging/src/k8s.io/apiserver/pkg/server/hooks.go#L238-L245
+
+Finally, we suggest starting just with builtin resources, as CRDs can be created at
+any time, making it harder to implement with unclear gain. If it appears to not be
+enough (or alternatively we realize that only a subset of builtin resource types is
+needed), this decision can easily be revisited in the future.
+
+#### Reject hanging watches
+
+To mitigate the problem of starving certain PriorityLevels, instead of having the
+watch hang and wait for watchcache initialization, we will simply reject it with
+`Too Many Requests` 429 http code.
+The reason for using this code is that it's already [properly handled] by our client
+libraries (e.g. reflector) and will not risk all of them suddenly falling back
+to list requests. Moreover, given that 429 is a long supported http code, we expect
+that other clients should also properly support it.
+
+[properly handled]: https://github.com/kubernetes/kubernetes/blob/41e706600aea7468f486150d951d3b8948ce89d5/staging/src/k8s.io/client-go/tools/cache/reflector.go#L910-L920
+
+#### Delegate get requests to etcd
+
+Since GET requests have bounded cost, we always estimate their weight the same
+no matter if they are served from etcd or from cache (the latency would be
+different in those cases though, so the actual seat-seconds cost too) and
+finally given we already did some work to process this request, we would
+simply delegate those requests to etcd.
+
+#### Adjust what lists are delegated to etcd
+
+It's tempting to reject all list requests with 429 the same way as watches.
+However, that would also have a negative consequences, e.g. by slowing down
+kube-apiserver initialization (kube-apiserver on startup need to initialize
+its own informers by loop-client, and by rejecting all lists until watchcache
+is initialized, we would effectively block that initialization).
+
+While we didn't decide to reject _all_ list requests with 429, we will start with
```

Comment:
could this be a "stick" to have people start using streaming lists?  Streaming lists always allowed, but regular lists rejected?

Reply:
No. Streaming lists will never be allowed here as they are served from watchcache - so if the watchcache is not initialized, we will not be able to serve them at all.